### PR TITLE
Clear Unknown Indices

### PR DESCRIPTION
**Pull request by bodo-hugo-barwich**
This implements the `--delete --all` functionality discussed at:
[Fix Corrupted Indices](https://github.com/metacpan/metacpan-api/issues/1048)
which is needed to close definitely the issue of Index Corruption
[Unexpected Index Corruption](https://github.com/metacpan/metacpan-docker/issues/47)
by
1) Preventing the Index Corruption
2) Recovering from Index Corruption with "_Clear All Indices_" option

Additionally it implements Exit with Error Code for unconfirmed Operations
as mentioned at:
[ElasticSearch Availabilty Check](https://github.com/metacpan/metacpan-api/pull/1045)
This is also part of the logic to **prevent** accidental Indices Corruption

To test the functionality it introduces several Mapping Script Tests.

These tests are required to check the "_Verify Mappings_" functionality developed at:
[Verify Indices Mappings](https://github.com/metacpan/metacpan-api/pull/1064)

### CODE REVIEW

**Comment by bodo-hugo-barwich** [comment]
The tests of the `--delete --all` logic show that this functionality is **save and should not be able to run in production** if my assumptions about the environment recognition are as documented in the related issue:
[recognize development environment](https://github.com/metacpan/metacpan-api/issues/1048#issuecomment-1186163526)
[flag development environment with environment variables](https://github.com/metacpan/metacpan-api/pull/1045#discussion_r767159095)
Still there is also a manual protection as requested at:
[manual confirmation required](https://github.com/metacpan/metacpan-api/pull/1045#discussion_r749563851)
which must throw an exception and exist the script

Of course the ultimate test will happen in the actual production environment where the script must fail on `--delete --all` and the prompt `ALL Indices will be deleted !!!` must not happen.

**Comment by mickeyn** [comment]
@bodo-hugo-barwich thanks for all this effort, but I'm a bit lost here, 
Can you please explain what is the purpose of all these big changes? they are all touching our very sensitive production indexing system but we don't have an idea of why this is all happening.

**Comment by bodo-hugo-barwich** [comment]
@mickeyn  thank you for taking a look at this development.

It was already discussed that this functionality **cannot be run in a production environment** and therefore it **cannot touch _sensitive production indexing system_**

I would always have preferred that you had taken part in your requirement discussion at:
[fix corrupted **development environment**](https://github.com/metacpan/metacpan-api/issues/1048#issuecomment-1068553982)
because at this state the development is already completed.

**Comment by bodo-hugo-barwich** [comment]
what is the state of this development ?

File: lib/MetaCPAN/Script/Mapping.pm [hunk 1]
```diff
@@ -180,6 +200,39 @@ sub delete_index {
...
+    }
+    else {
+        #Set System Error: 1 - EPERM - Operation not permitted
+        $self->exit_code(1);
+        $self->print_error("Operation not permitted!");
+        $self->handle_error(
+            "Operation not permitted in environment: $runtime_environment",
+            1 );
+    }
+}
```
Comment (oalders) [inline lib/MetaCPAN/Script/Mapping.pm:234]: Based on this, I think there are decent guards in place to prevent this from happening accidentally in production. @mickeyn what do you think?

File: lib/MetaCPAN/Script/Mapping.pm [hunk 2]
```diff
@@ -132,13 +141,18 @@ sub run {
     my $self = shift;
 
     if ( $self->await ) {
-        $self->delete_index if $self->arg_delete_index;
```
Comment (mickeyn) [inline lib/MetaCPAN/Script/Mapping.pm:135]: why the change of order? maybe we shouldn't allow a mix of these flags and have an if-else construct here to prevent that.

**Review by mickeyn — CHANGES_REQUESTED** [review]

File: lib/MetaCPAN/Script/Mapping.pm (hunk 1 above)
Comment (bodo-hugo-barwich) [reply lib/MetaCPAN/Script/Mapping.pm:273]: I also introduced a test to check that this protection is actually enforced:
```
$ docker-compose exec -T api_test prove -lr --jobs 2 t
2022/12/23 05:34:56 E mapping: Operation not permitted!
[2022/12/23 05:34:56] [catalyst] [ERROR] Operation not permitted!
2022/12/23 05:34:56 F mapping: Operation not permitted in environment: production
[2022/12/23 05:34:56] [catalyst] [FATAL] Operation not permitted in environment: production
```

File: lib/MetaCPAN/Script/Mapping.pm (hunk 2 above)
Comment (bodo-hugo-barwich) [reply lib/MetaCPAN/Script/Mapping.pm:150]: I have implemented the ` if-else ` check for all **editing operations** as requested.
I hope I understood well what you had in mind.
Please note that I also introduced **several tests** to check the editing operations.

Comment (mickeyn) [reply lib/MetaCPAN/Script/Mapping.pm:150]: thanks, I think that's better. if we ever need to mix flags we can change it accordingly, but that's a safer starting point.

**Review by mickeyn — DISMISSED** [review]

File: lib/MetaCPAN/Script/Mapping.pm (hunk 1 above)
Comment (bodo-hugo-barwich) [reply lib/MetaCPAN/Script/Mapping.pm:273]: It seems that the automated tests are not configured as development environment:
[automated _CirclCI_ test ` build-and-test (237)`](https://app.circleci.com/pipelines/github/metacpan/metacpan-api/229/workflows/b736e36e-3283-4d2b-b5e6-ea2e0092e2e6/jobs/237?invite=true#step-108-12)
```
2022/12/23 06:05:21 F mapping: Operation not permitted in environment: production
[2022/12/23 06:05:21] [catalyst] [FATAL] Operation not permitted in environment: production

        #   Failed test 'delete all succeeds'
        #   at t/lib/MetaCPAN/TestServer.pm line 434.

        #   Failed test 'Exit Code '0' - No Error'
        #   at t/lib/MetaCPAN/TestServer.pm line 435.
        #          got: '1'
        #     expected: '0'
        # Looks like you failed 2 tests of 2.
```

Which on the other hand shows that the protection works :)

**Comment by oalders** [comment]
@bodo-hugo-barwich this is failing in CI right now. Please see https://app.circleci.com/pipelines/github/metacpan/metacpan-api/229/workflows/b736e36e-3283-4d2b-b5e6-ea2e0092e2e6/jobs/237

**Comment by bodo-hugo-barwich** [comment]
Hi @oalders , yes, I saw that the tests are failing. 
The `create_index` operation requires that the index **does not exist** but it was created to **test** the `--delete --all` operation which did fail because the test environment **could not be recognized** as such:
```
2022/12/23 06:05:21 E mapping: Operation not permitted!
[2022/12/23 06:05:21] [catalyst] [ERROR] Operation not permitted!
2022/12/23 06:05:21 F mapping: Operation not permitted in environment: production
[2022/12/23 06:05:21] [catalyst] [FATAL] Operation not permitted in environment: production

        #   Failed test 'delete all succeeds'
        #   at t/lib/MetaCPAN/TestServer.pm line 434.

        #   Failed test 'Exit Code '0' - No Error'
        #   at t/lib/MetaCPAN/TestServer.pm line 435.
        #          got: '1'
        #     expected: '0'
        # Looks like you failed 2 tests of 2.

    #   Failed test 'delete all succeeds'
    #   at t/lib/MetaCPAN/TestServer.pm line 437.
 ```
@oalders Is there a way that the automated test environment can be marked as such? Perhaps setting the `MOJO_MODE` to an appropriate value?

**Comment by oalders** [comment]
@bodo-hugo-barwich I think that would be appropriate. Maybe we just need to set an env var in `.circleci/config.yml`? If it were part of this PR, that would be a helpful way to test it. https://circleci.com/docs/set-environment-variable/

**Comment by bodo-hugo-barwich** [comment]
The change in the Automated Testing Environment is part of the `metacpan-docker`.
[Set `MOJO_MODE` in Automated Testing Environment](https://github.com/metacpan/metacpan-docker/pull/95)
So, this will become an requirement for this merge to succeed.

**Comment by bodo-hugo-barwich** [comment]
With the `MOJO_MODE` setting in the _MetaCPAN_ API test the test suite should now succeed.
But the Automated Tests are not rerun since Friday.
https://app.circleci.com/pipelines/github/metacpan/metacpan-api/229/workflows/b736e36e-3283-4d2b-b5e6-ea2e0092e2e6/jobs/237
@oalders is it possible to trigger a **rerun**?

**Comment by oalders** [comment]
@bodo-hugo-barwich I have just triggered a re-run at CircleCI.

**Comment by bodo-hugo-barwich** [comment]
Inspecting the runtime environment during the test it turns out that the expected Environment Variables are never set:
https://app.circleci.com/pipelines/github/metacpan/metacpan-api/232/workflows/24008785-20e1-4fa8-9cb1-855daa386658/jobs/241
```
Use of uninitialized value $ENV{"PLACK_ENV"} in concatenation (.) or string at t/lib/MetaCPAN/TestServer.pm line 377.
test_delete_fails - PLACK_ENV: ''
Use of uninitialized value $ENV{"MOJO_MODE"} in concatenation (.) or string at t/lib/MetaCPAN/TestServer.pm line 378.
test_delete_fails - MOJO_MODE: ''
2022/12/30 13:28:50 E mapping: Operation not permitted!
[2022/12/30 13:28:50] [catalyst] [ERROR] Operation not permitted!
2022/12/30 13:28:50 F mapping: Operation not permitted in environment: production
[2022/12/30 13:28:50] [catalyst] [FATAL] Operation not permitted in environment: production
```
This can be understood from the fact that the test are run in the sandboxed _Docker_ environment. And this environment is defined in `docker-compose.yml`:
[`api_test` definition in `docker-compose.yml`](https://github.com/metacpan/metacpan-docker/blob/f8750035926cc7741714f4d885521fe8968276da/docker-compose.yml#L162-L163
)
```yaml
    env_file:
      - localapi_test.env
```      
Understanding this I found that I made two changes to the `localapi*.env` files:
```plain
$ git diff origin/no-94_launch_tests localapi_test.env
diff --git a/localapi_test.env b/localapi_test.env
index 4ea8bc2..1f49954 100644
--- a/localapi_test.env
+++ b/localapi_test.env
@@ -8,3 +8,4 @@ METACPAN_SERVER_CONFIG_LOCAL_SUFFIX=testing
 MINICPAN=/CPAN
 PERL_MM_USE_DEFAULT=1
 PERL_CARTON_PATH=/carton
+MOJO_MODE=testing
$ git diff origin/no-94_launch_tests localapi.env
diff --git a/localapi.env b/localapi.env
index f0a4aa4..97514a8 100644
--- a/localapi.env
+++ b/localapi.env
@@ -6,3 +6,4 @@ ES_TEST=elasticsearch_test:9200
 MINICPAN=/CPAN
 PERL_MM_USE_DEFAULT=1
 PERL_CARTON_PATH=/carton
+MOJO_MODE=development
```
So, I will need to revert the changes in the Automated Testing Environment and add those changes to the `metacpan-docker` project.

**Comment by bodo-hugo-barwich** [comment]
The changes in 
[`MOJO_MODE` for _Docker_ executions](https://github.com/metacpan/metacpan-docker/pull/96)
will introduce these settings in the _Docker_ environment.

**Comment by oalders** [comment]
This is only failing on a tidy test now, so we can merge and tidy in a different PR. Thanks for your patience @bodo-hugo-barwich and thank you for the reviews @mickeyn.

Happy New Year!

**Comment by mickeyn** [comment]
Cool. Thanks the work and follow up @bodo-hugo-barwich and @oalders .

Happy New Year! 😉

**Comment by bodo-hugo-barwich** [comment]
Thank you @oalders and @mickeyn for your trust! :smile: 

Happy New Year! 
And welcoming a new year of meaningful quality work! :smile: :tada:
